### DISCUSSION TITLE
Generating a value dependent on another field with a random offset

### DISCUSSION THREAD

**Discussion by pas-n** [instancio/instancio#1060] [Q&A] [answered]
Hi,

I am currently trying to create a `Model<TimeSpan>` object, that produces the same result if called using `withSeed`.

```java
record TimeSpan(Instant start, Instant end, AnotherObject other)

Instancio.of(createTimeSpanModel())
    .withSeed(123)
    .create();
```

The end should be between 1 and 10 days after the start.
I already tried `.assign` + `.as`, but it seems like there is no way to make the random object reproducible, since I don't have access to the random object in the `.as` method.

```java
Instancio.of(TimeSpan.class)
    .setModel(all(AnotherObject.class), ...)
    .generate(field(TimeSpan::start), gen -> gen.temporal().instant())
    .assign(valueOf(TimeSpan::start).to(field(TimeSpan::end).<Instant, Instant>as(ts -> ts.plus(???)))
```

(The other discussions I found about this topic also have that reproducibility problem)

For a simple TimeStamp object a custom generator might work, that's why I added the `other` field to illustrate that more complex objects might want to use other `.generate`, `.setModel` calls on the instancio object.

Did I miss something or is this currently not possible?

**Comment by armandino** [comment, marked as answer]
Your summary is accurate. Currently, the `as(Function)` method does not provide the `Random` object as an argument.

If you're working with a fixed seed, one option would be to generate the length of the period separately:

```java
final long seed  = 123;

Integer periodLength = Instancio.of(Integer.class)
    .withSeed(seed)
    .generate(allInts(), gen -> gen.ints().range(1, 10))
    .create();

TimeSpan timespan = Instancio.of(TimeSpan.class)
    .withSeed(seed)
    //... snip
    .assign(valueOf(TimeSpan::start)
                .to(field(TimeSpan::end))
                .as((Instant start) -> start.plus(periodLength, ChronoUnit.DAYS)))
    .create();
```

Another alternative is to use a custom generator as you mentioned. If you create a lot of `TimeSpan` objects in your tests that might be the better solution. Not sure if you were referring to this discussion (it has a somewhat similar use case):

- #856

**Reply by pas-n** [reply]
Hi, thanks for the clarification.

On thing I am struggling with, is using `Generators.temporal().instant()` or an already defined `Model` in my custom generator, since they don't expose a `create(Random)` method that I can call.
```java
Instancio.of(OtherModelClass.class).setModel(root(), createOtherModel()).create()
// or
Instancio.of(Instant.class).generate(root(), gen -> gen.temporal().instant().min(...).max(...)).create()
```
wouldn't work since I can't forward the Random object.

Do you have any advice on that, if something like that is possible?

**Reply by armandino** [reply]
No problem! Regarding you last question,  I'm mot sure if I'm following. Could you please provide a complete runnable example that I can try on my end?

**Reply by pas-n** [reply]
Sure, if I have a custom generator like this:

```java
public static class TimeSpanGenerator implements Generator<TimeSpan>
{
    @Override
    public TimeSpan generate(Random random)
    {
        // is there a better way to get an instance of this
        InstantGenerator instantGenerator = new InstantGenerator();

        Instant start = instantGenerator.generate(random);

        Instant end = instantGenerator
            .min(start)
            .max(start.plus(Duration.ofDays(10)))
            .generate(random);

        return new TimeSpan(start, end);
    }
}
```

I can use the InstantGenerator by instantiating a new instance, but I am not sure if that is the correct way?

And the other problem is, if the TimSpan object takes a third argument as I wrote in the original question,
but I already have a Model for that object, how would I use that in my generator?

```java
Model<AnotherObject> createAnotherObjectModel() {
    return Instancio.of(AnotherObject.class)
       .xyz(...)
       .toModel();
}

public static class TimeSpanGenerator implements Generator<TimeSpan>
{
    @Override
    public TimeSpan generate(Random random)
    {
        InstantGenerator instantGenerator = new InstantGenerator();

        Instant start = instantGenerator.generate(random);

        Instant end = instantGenerator
            .min(start)
            .max(start.plus(Duration.ofDays(10)))
            .generate(random);

        // Model<AnotherObject> has no method to create a new instance
        AnotherObject obj = createAnotherObjectModel().????();

        return new TimeSpan(start, end, obj);
    }
}
```

The following doesn't work because I can't supply my own Random object.
```java
Instancio.of(AnotherObject.class).setModel(root(), createAnotherObjectModel()).create()
```

**Reply by armandino** [reply]
@pas-n thanks for the additional info. I overlooked the fact that `TimeSpan` is a record. Unfortunately, a record created using a generator cannot be modified. Once a generator returns `new TimeSpan(start, end)`, the returned object cannot be customised using selectors since records are immutable. 

Probably the easiest option is to create a helper method to create a `Model<TimeSpan>`:

```java
private static Model<TimeSpan> timeSpanModel(final long seed) {
    Integer periodLength = Instancio.of(Integer.class)
            .withSeed(seed)
            .generate(root(), gen -> gen.ints().range(1, 10))
            .create();

    return Instancio.of(TimeSpan.class)
            .withSeed(seed)
            .generate(field(TimeSpan::start), gen -> gen.temporal().instant())
            .assign(valueOf(TimeSpan::start)
                    .to(field(TimeSpan::end))
                    .as((Instant start) -> start.plus(periodLength, ChronoUnit.DAYS)))
            .toModel();
}
```

then you can use the model to create `TimeSpan`:

```java
Model<TimeSpan> timeSpanModel = timeSpanModel(123);

TimeSpan timeSpan = Instancio.create(timeSpanModel);

// or
TimeSpan timeSpan = Instancio.of(timeSpanModel)
    .setModel(all(AnotherObject.class), anotherObjectModel)
    .create();
```

Hope this helps.
